specfile/prep.py

```diff
@@ -0,0 +1,303 @@
+        if self.options.P is not None:
+            return self.options.P
+        tokens = re.split(r"(\d+)", self._name, maxsplit=1)
+        if len(tokens) > 1:
+            return int(tokens[1])
+        return 0
+
+
+class AutosetupMacro(PrepMacro):
+    """Class that represents an %autosetup macro."""
+
+    def _setup_parser(self) -> None:
+        """Configures internal `ArgumentParser` for the options of the macro."""
+        self._parser.add_argument("-n", default="%{name}-%{version}")
+        self._parser.add_argument("-v", action="store_true")
+        self._parser.add_argument("-c", action="store_true")
+        self._parser.add_argument("-D", action="store_true")
+        self._parser.add_argument("-T", action="store_true")
+        self._parser.add_argument("-b", type=int)
+        self._parser.add_argument("-a", type=int)
+        self._parser.add_argument("-N", action="store_true")
+        self._parser.add_argument("-S", default="patch")
+        self._parser.add_argument("-p", type=int)
+
+
+class AutopatchMacro(PrepMacro):
+    """Class that represents an %autopatch macro."""
+
+    def _setup_parser(self) -> None:
+        """Configures internal `ArgumentParser` for the options of the macro."""
+        self._parser.add_argument("-v", action="store_true")
+        self._parser.add_argument("-p", type=int)
+        self._parser.add_argument("-m", type=int)
+        self._parser.add_argument("-M", type=int)
+        self._parser.add_argument("indices", type=int, nargs="*")
+
+
+class PrepMacros(collections.abc.Sequence):
+    """Class that represents a sequence of all %prep macros."""
+
+    def __init__(self, section: Section) -> None:
+        """
+        Constructs a `PrepMacros` object.
+
+        Args:
+            section: %prep section.
+
+        Returns:
+            Constructed instance of `PrepMacros` class.
+        """
+        self._section = section
+
+    def __repr__(self) -> str:
+        section = repr(self._section)
+        return f"PrepMacros({section})"
+
+    def __len__(self) -> int:
+        return len(self._get_items())
+
+    def _get_items(self) -> List[PrepMacro]:
+        """
+        Gets all supported %prep macros.
+
+        Returns:
+            List of instances of subclasses of PrepMacro.
+        """
+        comment_regex = re.compile(r"^\s*#.*$")
+        macro_regex = re.compile(
+            r"(?P<c>%{!?\?\w+:)?.*?"
+            r"(?P<m>%(setup|patch\d*|autopatch|autosetup))\s*"
+            r"(?P<o>.*?)(?(c)}|$)"
+        )
+        result: List[PrepMacro] = []
+        for i, line in enumerate(self._section):
+            if comment_regex.match(line):
+                continue
+            m = macro_regex.search(line)
+            if not m:
+                continue
+            name, options = m.group("m"), m.group("o")
+            macro: PrepMacro
+            if name == "%setup":
+                macro = SetupMacro(i, name)
+                macro._parse_options(options)
```

Comment:
Can't we take these _parse_options and append call out of all the ifs and have it here once for all the branches? The only exception will be the `%patch` special-case.